lib/seattleflu/id3c/cli/command/etl/redcap.py

```diff
@@ -44,3 +69,628 @@ def normalize_net_id(net_id: str=None) -> Optional[str]:
+        )
+    }
+
+
+def _create_patient(sex: str, preferred_language: str, first_name: str, last_name: str,
+        birth_date: str, zipcode: str, record: REDCapRecord,
+        system_identifier: str, unique_identifier: str = None) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses demographics to create the patient identifier unless
+    a *unique_identifier* is provided.
+    """
+    gender = map_sex(sex)
+
+    language_codeable_concept = create_codeable_concept(
+        system = 'urn:ietf:bcp:47',
+        code = preferred_language
+    )
+
+    communication = [{
+        'language' : language_codeable_concept,
+        'preferred': True
+    }]
+
+    patient_id = None
+    if unique_identifier:
+        patient_id = generate_hash(unique_identifier)
+    else:
+        patient_id = generate_patient_hash(
+                names       = (first_name, last_name),
+                gender      = gender,
+                birth_date  = birth_date,
+                postal_code = zipcode)
+
+    if not patient_id:
+        # Some piece of information was missing, so we couldn't generate a
+        # hash.  Fallback to treating this individual as always unique by using
+        # the REDCap record id.
+        patient_id = generate_hash(f"{record.project.base_url}{record.project.id}/{record.id}")
+
+    LOG.debug(f"Generated individual identifier {patient_id}")
+
+    patient_identifier = create_identifier(f"{system_identifier}/individual", patient_id)
+    patient_resource = create_patient_resource([patient_identifier], gender, communication)
+
+    return create_entry_and_reference(patient_resource, "Patient")
+
+
+def create_patient_using_demographics(sex: str, preferred_language: str, first_name: str, last_name: str,
+        birth_date: str, zipcode: str, record: REDCapRecord, system_identifier: str) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses demographics to create the patient identifier
+    """
+    return _create_patient(
+        sex = sex,
+        preferred_language = preferred_language,
+        first_name = first_name,
+        last_name = last_name,
+        birth_date = birth_date,
+        zipcode = zipcode,
+        record = record,
+        system_identifier = system_identifier,
+        unique_identifier = None)
+
+
+def create_patient_using_unique_identifier(sex: str, preferred_language: str, unique_identifier: str,
+    record: REDCapRecord, system_identifier: str) -> tuple:
+    """
+    Returns a FHIR Patient resource entry and reference.
+    Uses a unique identifier to create the patient identifier
+    """
+    return _create_patient(
+        sex = sex,
+        preferred_language = preferred_language,
+        first_name = None,
+        last_name = None,
+        birth_date = None,
+        zipcode = None,
+        record = record,
+        system_identifier = system_identifier,
+        unique_identifier = unique_identifier)
```

Comment:
I'm not sure I understand why this needs to be two separate functions. With the `unique_identifier` being an optional argument, you can call the inner function directly without the two wrappers. The inner `_create_patient` function already does the right thing of checking whether or not `unique_identifier` is available.

Reply:
It's for better usability. I wanted explicit functions for creating a patient with a unique identifier or with demographics. This prevents the caller from having to pass in None values for first_name, last_name, birth_date, and zip_code if they want to use a unique identifier.

Reply:
Hmm, then why not make the demographic fields optional as well. So `create_patient` can do the right thing without the caller having to pass `None` values?

Reply:
All the optional parameters have to come at the end of the function signature, right? I didn't really like sticking those all on the far right.